advection_fv4/interface.py

```diff
@@ -0,0 +1,302 @@
+def states(a, qx, qy, ng, idir):
+    r"""
+    Predict the cell-centered state to the edges in one-dimension using the
+    reconstructed, limited slopes. We use a fourth-order Godunov method.
+
+    Our convention here is that:
+
+        ``al[i,j]``   will be :math:`al_{i-1/2,j}`,
+
+        ``al[i+1,j]`` will be :math:`al_{i+1/2,j}`.
+
+    Parameters
+    ----------
+    a : ndarray
+        The cell-centered state.
+    qx, qy : int
+        The dimensions of `a`.
+    ng : int
+        The number of ghost cells
+    idir : int
+        Are we predicting to the edges in the x-direction (1) or y-direction (2)?
+
+    Returns
+    -------
+    out : ndarray, ndarray
+        The state predicted to the left and right edges.
+    """
+
+    al = np.zeros((qx, qy))
+    ar = np.zeros((qx, qy))
+
+    a_int = np.zeros((qx, qy))
+    dafm = np.zeros((qx, qy))
+    dafp = np.zeros((qx, qy))
+    d2af = np.zeros((qx, qy))
+    d2ac = np.zeros((qx, qy))
+    d3a = np.zeros((qx, qy))
+
+    C2 = 1.25
+    C3 = 0.1
+
+    nx = qx - 2 * ng
+    ny = qy - 2 * ng
+    ilo = ng
+    ihi = ng + nx
+    jlo = ng
+    jhi = ng + ny
+
+    # we need interface values on all faces of the domain
+    if (idir == 1):
+
+        for j in range(jlo - 1, jhi + 1):
+            for i in range(ilo - 2, ihi + 3):
+
+                # interpolate to the edges
+                a_int[i, j] = (7.0 / 12.0) * (a[i - 1, j] + a[i, j]) - \
+                    (1.0 / 12.0) * (a[i - 2, j] + a[i + 1, j])
+
+                al[i, j] = a_int[i, j]
+                ar[i, j] = a_int[i, j]
+
+        for j in range(jlo - 1, jhi + 1):
+            for i in range(ilo - 2, ihi + 3):
+                # these live on cell-centers
+                dafm[i, j] = a[i, j] - a_int[i, j]
+                dafp[i, j] = a_int[i + 1, j] - a[i, j]
+
+                # these live on cell-centers
+                d2af[i, j] = 6.0 * (a_int[i, j] - 2.0 *
+                                    a[i, j] + a_int[i + 1, j])
+
+        for j in range(jlo - 1, jhi + 1):
+            for i in range(ilo - 3, ihi + 3):
+                d2ac[i, j] = a[i - 1, j] - 2.0 * a[i, j] + a[i + 1, j]
+
+        for j in range(jlo - 1, jhi + 1):
+            for i in range(ilo - 2, ihi + 3):
+                # this lives on the interface
+                d3a[i, j] = d2ac[i, j] - d2ac[i - 1, j]
+
+        # this is a look over cell centers, affecting
+        # i-1/2,R and i+1/2,L
+        for j in range(jlo - 1, jhi + 1):
+            for i in range(ilo - 1, ihi + 1):
+
+                # limit? MC Eq. 24 and 25
+                if (dafm[i, j] * dafp[i, j] <= 0.0 or
+                        (a[i, j] - a[i - 2, j]) * (a[i + 2, j] - a[i, j]) <= 0.0):
+
+                    # we are at an extrema
+
+                    s = np.copysign(1.0, d2ac[i, j])
+                    if (s == np.copysign(1.0, d2ac[i - 1, j]) and s == np.copysign(1.0, d2ac[i + 1, j]) and
+                            s == np.copysign(1.0, d2af[i, j])):
```

Comment:
copysign? or sign?

Reply:
[`np.copysign`](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.copysign.html) seems to have the same behaviour as Fortran's [`sign`](https://gcc.gnu.org/onlinedocs/gfortran/SIGN.html)